File: go/host/peer_manager.go
package host

import (
	"errors"
	"sync"
	"time"

	"github.com/vibing/zgrnet/noise"
)

// pendingHandshake tracks an in-progress handshake.
type pendingHandshake struct {
	peer      *Peer
	hsState   *noise.HandshakeState
	localIdx  uint32
	done      chan error
	createdAt time.Time
}

// PeerManager manages all peers and their connections.
// It provides routing based on session index and public key lookup.
type PeerManager struct {
	mu sync.RWMutex

	// Peer lookup
	byPubkey map[noise.PublicKey]*Peer
	byIndex  map[uint32]*Peer // session index -> peer

	// Pending handshakes (by local index)
	pending map[uint32]*pendingHandshake

	// Local identity
	localKey *noise.KeyPair

	// Transport (shared for all peers)
	transport noise.Transport
}

// NewPeerManager creates a new peer manager.
func NewPeerManager(localKey *noise.KeyPair, transport noise.Transport) *PeerManager {
	return &PeerManager{
		byPubkey:  make(map[noise.PublicKey]*Peer),
		byIndex:   make(map[uint32]*Peer),
		pending:   make(map[uint32]*pendingHandshake),
		localKey:  localKey,
		transport: transport,
	}
}

// AddPeer registers a new peer.
// If a peer with the same public key already exists, it returns an error.
func (pm *PeerManager) AddPeer(peer *Peer) error {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	pk := peer.PublicKey()
	if _, exists := pm.byPubkey[pk]; exists {
		return ErrPeerExists
	}

	pm.byPubkey[pk] = peer
	return nil
}

// RemovePeer removes a peer by public key.
func (pm *PeerManager) RemovePeer(pk noise.PublicKey) {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	peer, exists := pm.byPubkey[pk]
	if !exists {
		return
	}

	// Remove from index mapping
	if session := peer.Session(); session != nil {
		delete(pm.byIndex, session.LocalIndex())
	}

	// Clear the session
	peer.ClearSession()

	delete(pm.byPubkey, pk)
}

// GetPeer returns a peer by public key.
func (pm *PeerManager) GetPeer(pk noise.PublicKey) *Peer {
	pm.mu.RLock()
	defer pm.mu.RUnlock()
	return pm.byPubkey[pk]
}

// GetPeerByIndex returns a peer by session index.
func (pm *PeerManager) GetPeerByIndex(index uint32) *Peer {
	pm.mu.RLock()
	defer pm.mu.RUnlock()
	return pm.byIndex[index]
}

// ListPeers returns all peers.
func (pm *PeerManager) ListPeers() []*Peer {
	pm.mu.RLock()
	defer pm.mu.RUnlock()

	peers := make([]*Peer, 0, len(pm.byPubkey))
	for _, peer := range pm.byPubkey {
		peers = append(peers, peer)
	}
	return peers
}

// Count returns the number of peers.
func (pm *PeerManager) Count() int {
	pm.mu.RLock()
	defer pm.mu.RUnlock()
	return len(pm.byPubkey)
}

// Dial initiates a connection to a peer.
// The peer must already be registered with AddPeer.
// This method blocks until the handshake completes or times out.
func (pm *PeerManager) Dial(pk noise.PublicKey) error {
	return pm.DialWithTimeout(pk, 10*time.Second)
}

// DialWithTimeout initiates a connection with a custom timeout.
func (pm *PeerManager) DialWithTimeout(pk noise.PublicKey, timeout time.Duration) error {
	peer := pm.GetPeer(pk)
	if peer == nil {
		return ErrPeerNotFound
	}

	if peer.IsEstablished() {
		return nil // Already connected
	}

	endpoint := peer.Endpoint()
	if endpoint == nil {
		return ErrNoEndpoint
	}

	// Start async handshake
	done, err := pm.dialAsync(peer, endpoint)
	if err != nil {
		return err
	}

	// Wait for completion or timeout
	select {
	case err := <-done:
		return err
	case <-time.After(timeout):
		// Clean up pending handshake
		pm.mu.Lock()
		for idx, p := range pm.pending {
			if p.peer == peer {
				delete(pm.pending, idx)
				break
			}
		}
		pm.mu.Unlock()
		peer.SetState(PeerStateFailed)
		return ErrHandshakeTimeout
	}
}

// dialAsync starts a handshake without blocking.
// Returns a channel that receives nil on success or an error on failure.
func (pm *PeerManager) dialAsync(peer *Peer, endpoint noise.Addr) (chan error, error) {
	peer.SetState(PeerStateConnecting)

	// Generate local index
	localIdx, err := noise.GenerateIndex()
	if err != nil {
		peer.SetState(PeerStateFailed)
		return nil, err
	}

	// Create handshake state (IK pattern - we know remote's public key)
	remotePK := peer.publicKey
	hs, err := noise.NewHandshakeState(noise.Config{
		Pattern:      noise.PatternIK,
		Initiator:    true,
		LocalStatic:  pm.localKey,
		RemoteStatic: &remotePK,
	})
	if err != nil {
		peer.SetState(PeerStateFailed)
		return nil, err
	}

	// Generate handshake initiation message
	msg1, err := hs.WriteMessage(nil)
	if err != nil {
		peer.SetState(PeerStateFailed)
		return nil, err
	}

	// Build wire message
	wireMsg := noise.BuildHandshakeInit(localIdx, hs.LocalEphemeral(), msg1[noise.KeySize:])

	// Register pending handshake before sending (to handle fast responses)
	done := make(chan error, 1)
	pm.mu.Lock()
	pm.pending[localIdx] = &pendingHandshake{
		peer:      peer,
		hsState:   hs,
		localIdx:  localIdx,
		done:      done,
		createdAt: time.Now(),
	}
	pm.mu.Unlock()

	// Send handshake initiation
	if err := pm.transport.SendTo(wireMsg, endpoint); err != nil {
		pm.mu.Lock()
		delete(pm.pending, localIdx)
		pm.mu.Unlock()
		peer.SetState(PeerStateFailed)
		return nil, err
	}

	return done, nil
}

// HandleHandshakeResp processes an incoming handshake response.
// This is called by the receive loop when a response is received.
func (pm *PeerManager) HandleHandshakeResp(data []byte, from noise.Addr) error {
	resp, err := noise.ParseHandshakeResp(data)
	if err != nil {
		return err
	}

	pm.mu.Lock()
	pending, exists := pm.pending[resp.ReceiverIndex]
	if !exists {
		pm.mu.Unlock()
		return ErrNoPendingHandshake
	}
	delete(pm.pending, resp.ReceiverIndex)
	pm.mu.Unlock()

	peer := pending.peer
	hs := pending.hsState

	// Reconstruct the noise message and process
	noiseMsg := make([]byte, noise.KeySize+16)
	copy(noiseMsg[:noise.KeySize], resp.Ephemeral[:])
	copy(noiseMsg[noise.KeySize:], resp.Empty)

	if _, err := hs.ReadMessage(noiseMsg); err != nil {
		peer.SetState(PeerStateFailed)
		pending.done <- err
		return err
	}

	// Get transport keys
	sendCS, recvCS, err := hs.Split()
	if err != nil {
		peer.SetState(PeerStateFailed)
		pending.done <- err
		return err
	}

	// Create session
	session, err := noise.NewSession(noise.SessionConfig{
		LocalIndex:  pending.localIdx,
		RemoteIndex: resp.SenderIndex,
		SendKey:     sendCS.Key(),
		RecvKey:     recvCS.Key(),
		RemotePK:    peer.publicKey,
	})
	if err != nil {
		peer.SetState(PeerStateFailed)
		pending.done <- err
		return err
	}

	// Register session
	pm.mu.Lock()
	peer.SetSession(session)
	pm.byIndex[session.LocalIndex()] = peer
	pm.mu.Unlock()

	// Update endpoint (roaming support)
	peer.SetEndpoint(from)

	// Signal success
	pending.done <- nil
	return nil
}

// HandleHandshakeInit processes an incoming handshake initiation.
// If the peer is known, it accepts the handshake and establishes a session.
// If the peer is unknown, it creates a new peer (if allowUnknown is true).
func (pm *PeerManager) HandleHandshakeInit(data []byte, from noise.Addr, allowUnknown bool) error {
	msg, err := noise.ParseHandshakeInit(data)
	if err != nil {
		return err
	}

	// Create a connection to process the handshake
	conn, err := noise.NewConn(noise.ConnConfig{
		LocalKey:   pm.localKey,
		Transport:  pm.transport,
		RemoteAddr: from,
	})
	if err != nil {
		return err
	}

	// Process the handshake (this extracts the remote public key)
	resp, err := conn.Accept(msg)
	if err != nil {
		return err
	}

	remotePK := conn.RemotePublicKey()
	session := conn.Session()
	if session == nil {
		return ErrHandshakeFailed
	}

	pm.mu.Lock()
	defer pm.mu.Unlock()

	// Check if we know this peer
	peer, exists := pm.byPubkey[remotePK]
	if !exists {
		if !allowUnknown {
			return ErrUnknownPeer
		}
		// Create a new peer for this connection
		peer = NewPeer(PeerConfig{
			PublicKey: remotePK,
			Endpoint:  from,
		})
		pm.byPubkey[remotePK] = peer
	}

	// Remove old session index mapping if exists
	if oldSession := peer.Session(); oldSession != nil {
		delete(pm.byIndex, oldSession.LocalIndex())
	}

	// Update peer with new session
	peer.SetSession(session)
	peer.SetEndpoint(from) // Update endpoint (roaming)
	pm.byIndex[session.LocalIndex()] = peer

	// Send the response
	return pm.transport.SendTo(resp, from)
}

// HandleTransport processes an incoming transport message.
// Returns the peer, protocol, and payload if successful.
func (pm *PeerManager) HandleTransport(data []byte, from noise.Addr) (*Peer, byte, []byte, error) {
	msg, err := noise.ParseTransportMessage(data)
	if err != nil {
		return nil, 0, nil, err
	}

	pm.mu.RLock()
	peer := pm.byIndex[msg.ReceiverIndex]
	pm.mu.RUnlock()

	if peer == nil {
		return nil, 0, nil, ErrSessionNotFound
	}

	session := peer.Session()
	if session == nil {
		return nil, 0, nil, ErrSessionNotFound
	}

	// Decrypt
	plaintext, err := session.Decrypt(msg.Ciphertext, msg.Counter)
	if err != nil {
		return nil, 0, nil, err
	}

	// Decode protocol and payload
	protocol, payload, err := noise.DecodePayload(plaintext)
	if err != nil {
		return nil, 0, nil, err
	}

	// Update stats and roaming
	peer.AddRxBytes(uint64(len(data)))
	peer.UpdateActivity()

	// Roaming: update endpoint if it changed (only after successful decrypt)
	currentEndpoint := peer.Endpoint()
	if currentEndpoint == nil || currentEndpoint.String() != from.String() {
		peer.SetEndpoint(from)
	}

	return peer, protocol, payload, nil
}

// Send sends an encrypted message to a peer.
func (pm *PeerManager) Send(pk noise.PublicKey, protocol byte, payload []byte) error {
	peer := pm.GetPeer(pk)
	if peer == nil {
		return ErrPeerNotFound
	}

	if !peer.IsEstablished() {
		return ErrNotEstablished
	}

	session := peer.Session()
	if session == nil {
		return ErrNotEstablished
	}

	endpoint := peer.Endpoint()
	if endpoint == nil {
		return ErrNoEndpoint
	}

	// Encode payload with protocol
	plaintext := noise.EncodePayload(protocol, payload)

	// Encrypt
	ciphertext, counter, err := session.Encrypt(plaintext)
	if err != nil {
		return err
	}

	// Build and send message
	msg := noise.BuildTransportMessage(session.RemoteIndex(), counter, ciphertext)

	peer.AddTxBytes(uint64(len(msg)))
	peer.UpdateActivity()

	return pm.transport.SendTo(msg, endpoint)
}

// ExpireStale removes peers with expired sessions.
// Returns the number of peers removed.
func (pm *PeerManager) ExpireStale() int {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	var expired []*Peer
	for _, peer := range pm.byPubkey {
		if peer.IsExpired() {
			expired = append(expired, peer)
		}
	}

	for _, peer := range expired {
		if session := peer.Session(); session != nil {
			delete(pm.byIndex, session.LocalIndex())
		}
		peer.ClearSession()
	}

	return len(expired)
}

// StartMaintenance starts a background goroutine for periodic maintenance.
// Returns a stop function.
func (pm *PeerManager) StartMaintenance(interval time.Duration) func() {
	stop := make(chan struct{})
	done := make(chan struct{})

	go func() {
		defer close(done)
		ticker := time.NewTicker(interval)
		defer ticker.Stop()

		for {
			select {
			case <-ticker.C:
				pm.ExpireStale()
			case <-stop:
				return
			}
		}
	}()

	return func() {
		close(stop)
		<-done
	}
}

// ExpirePendingHandshakes removes handshakes that have been pending too long.
// Returns the number of handshakes expired.
func (pm *PeerManager) ExpirePendingHandshakes(maxAge time.Duration) int {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	var expired []uint32
	now := time.Now()
	for idx, p := range pm.pending {
		if now.Sub(p.createdAt) > maxAge {
			expired = append(expired, idx)
		}
	}

	for _, idx := range expired {
		if p, ok := pm.pending[idx]; ok {
			p.peer.SetState(PeerStateFailed)
			select {
			case p.done <- ErrHandshakeTimeout:
			default:
			}
			delete(pm.pending, idx)
		}
	}

	return len(expired)
}

// Clear removes all peers and pending handshakes.
func (pm *PeerManager) Clear() {
	pm.mu.Lock()
	defer pm.mu.Unlock()

	// Cancel all pending handshakes
	for _, p := range pm.pending {
		select {
		case p.done <- ErrHostClosed:
		default:
		}
	}
	pm.pending = make(map[uint32]*pendingHandshake)

	for _, peer := range pm.byPubkey {
		peer.ClearSession()
	}

	pm.byPubkey = make(map[noise.PublicKey]*Peer)
	pm.byIndex = make(map[uint32]*Peer)
}

// PeerManager errors.
var (
	ErrPeerExists         = errors.New("host: peer already exists")
	ErrPeerNotFound       = errors.New("host: peer not found")
	ErrNoEndpoint         = errors.New("host: peer has no endpoint")
	ErrHandshakeFailed    = errors.New("host: handshake failed")
	ErrHandshakeTimeout   = errors.New("host: handshake timed out")
	ErrNoPendingHandshake = errors.New("host: no pending handshake for index")
	ErrUnknownPeer        = errors.New("host: unknown peer")
	ErrSessionNotFound    = errors.New("host: session not found")
	ErrNotEstablished     = errors.New("host: connection not established")
)

File: zig/src/peer_manager.zig
//! Peer manager for managing connections to multiple peers.

const std = @import("std");
const Mutex = std.Thread.Mutex;
const Allocator = std.mem.Allocator;

const keypair = @import("keypair.zig");
const session_mod = @import("session.zig");
const transport_mod = @import("transport.zig");
const message_mod = @import("message.zig");
const peer_mod = @import("peer.zig");
const handshake_mod = @import("handshake.zig");

pub const Key = keypair.Key;
pub const KeyPair = keypair.KeyPair;
pub const Session = session_mod.Session;
pub const Transport = transport_mod.Transport;
pub const Addr = transport_mod.Addr;
pub const Peer = peer_mod.Peer;
pub const PeerConfig = peer_mod.PeerConfig;
pub const PeerState = peer_mod.PeerState;

/// Peer manager errors.
pub const PeerManagerError = error{
    PeerExists,
    PeerNotFound,
    NoEndpoint,
    HandshakeFailed,
    HandshakeTimeout,
    NoPendingHandshake,
    UnknownPeer,
    SessionNotFound,
    NotEstablished,
    HostClosed,
    TransportError,
    SessionError,
    MessageError,
    OutOfMemory,
};

/// Pending handshake state.
const PendingHandshake = struct {
    peer: *Peer,
    hs_state: handshake_mod.HandshakeState,
    local_idx: u32,
    created_at: i128,
};

/// Manages all peers and their connections.
pub const PeerManager = struct {
    allocator: Allocator,
    mutex: Mutex = .{},
    by_pubkey: std.AutoHashMap(Key, *Peer),
    by_index: std.AutoHashMap(u32, *Peer),
    pending: std.AutoHashMap(u32, PendingHandshake),
    local_key: KeyPair,
    transport: Transport,

    /// Creates a new peer manager.
    pub fn init(allocator: Allocator, local_key: KeyPair, transport: Transport) PeerManager {
        return .{
            .allocator = allocator,
            .by_pubkey = std.AutoHashMap(Key, *Peer).init(allocator),
            .by_index = std.AutoHashMap(u32, *Peer).init(allocator),
            .pending = std.AutoHashMap(u32, PendingHandshake).init(allocator),
            .local_key = local_key,
            .transport = transport,
        };
    }

    /// Deinitializes the peer manager and frees all peers.
    pub fn deinit(self: *PeerManager) void {
        self.mutex.lock();
        defer self.mutex.unlock();

        // Clear pending handshakes
        self.pending.clearAndFree();

        var it = self.by_pubkey.valueIterator();
        while (it.next()) |peer_ptr| {
            peer_ptr.*.clearSession();
            self.allocator.destroy(peer_ptr.*);
        }

        self.by_pubkey.deinit();
        self.by_index.deinit();
    }

    /// Adds a new peer.
    pub fn addPeer(self: *PeerManager, peer: *Peer) PeerManagerError!void {
        self.mutex.lock();
        defer self.mutex.unlock();

        const pk = peer.public_key;
        if (self.by_pubkey.contains(pk)) {
            return PeerManagerError.PeerExists;
        }

        self.by_pubkey.put(pk, peer) catch return PeerManagerError.OutOfMemory;
    }

    /// Removes a peer by public key.
    pub fn removePeer(self: *PeerManager, pk: Key) void {
        self.mutex.lock();
        defer self.mutex.unlock();

        if (self.by_pubkey.fetchRemove(pk)) |removed| {
            const peer = removed.value;
            // Remove from index mapping
            if (peer.getSession()) |session| {
                _ = self.by_index.remove(session.local_index);
            }
            peer.clearSession();
            self.allocator.destroy(peer);
        }
    }

    /// Gets a peer by public key.
    pub fn getPeer(self: *PeerManager, pk: Key) ?*Peer {
        self.mutex.lock();
        defer self.mutex.unlock();
        return self.by_pubkey.get(pk);
    }

    /// Gets a peer by session index.
    pub fn getPeerByIndex(self: *PeerManager, index: u32) ?*Peer {
        self.mutex.lock();
        defer self.mutex.unlock();
        return self.by_index.get(index);
    }

    /// Returns the number of peers.
    pub fn count(self: *PeerManager) usize {
        self.mutex.lock();
        defer self.mutex.unlock();
        return self.by_pubkey.count();
    }

    /// Sends a message to a peer.
    pub fn send(self: *PeerManager, pk: Key, protocol: message_mod.Protocol, payload: []const u8) PeerManagerError!void {
        const peer = self.getPeer(pk) orelse return PeerManagerError.PeerNotFound;

        if (!peer.isEstablished()) {
            return PeerManagerError.NotEstablished;
        }

        const endpoint = peer.getEndpoint() orelse return PeerManagerError.NoEndpoint;

        // Get session and encrypt
        const session = peer.getSession() orelse return PeerManagerError.NotEstablished;

        // Encode payload with protocol
        const plaintext = self.allocator.alloc(u8, 1 + payload.len) catch return PeerManagerError.OutOfMemory;
        defer self.allocator.free(plaintext);
        plaintext[0] = @intFromEnum(protocol);
        @memcpy(plaintext[1..], payload);

        // Encrypt
        const ciphertext = self.allocator.alloc(u8, plaintext.len + session_mod.tag_size) catch return PeerManagerError.OutOfMemory;
        defer self.allocator.free(ciphertext);

        const counter = session.encrypt(plaintext, ciphertext) catch return PeerManagerError.SessionError;

        // Build and send message
        const msg = message_mod.buildTransportMessage(self.allocator, session.remoteIndex(), counter, ciphertext) catch return PeerManagerError.OutOfMemory;
        defer self.allocator.free(msg);

        peer.addTxBytes(msg.len);
        peer.updateActivity();

        self.transport.sendTo(msg, endpoint) catch return PeerManagerError.TransportError;
    }

    /// Handles an incoming transport message.
    pub fn handleTransport(self: *PeerManager, data: []const u8, from: Addr) PeerManagerError!struct { peer: *Peer, protocol: message_mod.Protocol, payload: []const u8 } {
        const msg = message_mod.parseTransportMessage(data) catch return PeerManagerError.MessageError;

        const peer = self.getPeerByIndex(msg.receiver_index) orelse return PeerManagerError.SessionNotFound;

        const session = peer.getSession() orelse return PeerManagerError.SessionNotFound;

        // Decrypt
        const plaintext_len = msg.ciphertext.len - session_mod.tag_size;
        const plaintext = self.allocator.alloc(u8, plaintext_len) catch return PeerManagerError.OutOfMemory;
        errdefer self.allocator.free(plaintext);

        _ = session.decrypt(msg.ciphertext, msg.counter, plaintext) catch return PeerManagerError.SessionError;

        // Decode protocol and payload
        const decoded = message_mod.decodePayload(plaintext) catch return PeerManagerError.MessageError;

        // Update stats and roaming
        peer.addRxBytes(data.len);
        peer.updateActivity();

        // Roaming: update endpoint if changed
        const current = peer.getEndpoint();
        if (current == null) {
            peer.setEndpoint(from);
        }

        return .{
            .peer = peer,
            .protocol = decoded.protocol,
            .payload = decoded.payload,
        };
    }

    /// Expires stale peers.
    pub fn expireStale(self: *PeerManager) usize {
        self.mutex.lock();
        defer self.mutex.unlock();

        var expired = std.ArrayListUnmanaged(Key){};
        defer expired.deinit(self.allocator);

        var it = self.by_pubkey.iterator();
        while (it.next()) |entry| {
            if (entry.value_ptr.*.isExpired()) {
                expired.append(self.allocator, entry.key_ptr.*) catch continue;
            }
        }

        for (expired.items) |pk| {
            if (self.by_pubkey.get(pk)) |peer| {
                if (peer.getSession()) |session| {
                    _ = self.by_index.remove(session.local_index);
                }
                peer.clearSession();
            }
        }

        return expired.items.len;
    }

    /// Expires pending handshakes that are too old.
    pub fn expirePendingHandshakes(self: *PeerManager, max_age_ns: i128) usize {
        self.mutex.lock();
        defer self.mutex.unlock();

        const now = std.time.nanoTimestamp();
        var expired = std.ArrayListUnmanaged(u32){};
        defer expired.deinit(self.allocator);

        var it = self.pending.iterator();
        while (it.next()) |entry| {
            if (now - entry.value_ptr.created_at > max_age_ns) {
                expired.append(self.allocator, entry.key_ptr.*) catch continue;
            }
        }

        for (expired.items) |idx| {
            if (self.pending.fetchRemove(idx)) |removed| {
                removed.value.peer.setState(.failed);
            }
        }

        return expired.items.len;
    }

    /// Clears all peers and pending handshakes.
    pub fn clear(self: *PeerManager) void {
        self.mutex.lock();
        defer self.mutex.unlock();

        // Clear pending handshakes
        var pending_it = self.pending.valueIterator();
        while (pending_it.next()) |p| {
            p.peer.setState(.failed);
        }
        self.pending.clearAndFree();

        var it = self.by_pubkey.valueIterator();
        while (it.next()) |peer_ptr| {
            peer_ptr.*.clearSession();
            self.allocator.destroy(peer_ptr.*);
        }

        self.by_pubkey.clearAndFree();
        self.by_index.clearAndFree();
    }

    /// Registers a session for a peer.
    pub fn registerSession(self: *PeerManager, pk: Key, session: *Session) void {
        self.mutex.lock();
        defer self.mutex.unlock();

        if (self.by_pubkey.get(pk)) |peer| {
            // Remove old index mapping
            if (peer.getSession()) |old_session| {
                _ = self.by_index.remove(old_session.local_index);
            }

            peer.setSession(session);
            self.by_index.put(session.local_index, peer) catch {};
        }
    }

    /// Registers a peer by its local index.
    pub fn registerIndex(self: *PeerManager, local_index: u32, peer: *Peer) void {
        self.mutex.lock();
        defer self.mutex.unlock();
        self.by_index.put(local_index, peer) catch {};
    }

    /// Handles an incoming handshake initiation (responder side).
    pub fn handleHandshakeInit(self: *PeerManager, data: []const u8, from: Addr, allow_unknown: bool) PeerManagerError!void {
        const hs_init = message_mod.parseHandshakeInit(data) catch return PeerManagerError.MessageError;

        // Create responder handshake state
        var hs = handshake_mod.HandshakeState.init(.{
            .pattern = .ik,
            .initiator = false,
            .local_static = self.local_key,
            .remote_static = null, // Will be learned from message
        }) catch return PeerManagerError.HandshakeFailed;

        // Reconstruct noise message
        var noise_msg: [keypair.key_size + keypair.key_size + 16 + 16]u8 = undefined;
        @memcpy(noise_msg[0..keypair.key_size], &hs_init.ephemeral);
        @memcpy(noise_msg[keypair.key_size..], hs_init.static_encrypted);

        // Process handshake
        var payload_buf: [64]u8 = undefined;
        _ = hs.readMessage(&noise_msg, &payload_buf) catch return PeerManagerError.HandshakeFailed;

        // Get initiator's static key
        const initiator_pk = hs.remoteStatic() orelse return PeerManagerError.HandshakeFailed;
        if (initiator_pk.isZero()) {
            return PeerManagerError.HandshakeFailed;
        }

        // Find or create peer
        var peer: *Peer = undefined;
        {
            self.mutex.lock();
            defer self.mutex.unlock();

            if (self.by_pubkey.get(initiator_pk)) |p| {
                peer = p;
            } else if (allow_unknown) {
                peer = self.allocator.create(Peer) catch return PeerManagerError.OutOfMemory;
                peer.* = Peer.init(self.allocator, .{
                    .public_key = initiator_pk,
                    .endpoint = from,
                });
                self.by_pubkey.put(initiator_pk, peer) catch {
                    self.allocator.destroy(peer);
                    return PeerManagerError.OutOfMemory;
                };
            } else {
                return PeerManagerError.UnknownPeer;
            }
        }

        // Generate our index
        const local_idx = session_mod.generateIndex();

        // Generate response
        var msg2_buf: [256]u8 = undefined;
        const msg2 = hs.writeMessage(&.{}, &msg2_buf) catch return PeerManagerError.HandshakeFailed;

        // Get ephemeral
        const ephemeral = hs.localEphemeral() orelse return PeerManagerError.HandshakeFailed;

        // Build wire message
        var wire_buf: [128]u8 = undefined;
        const wire_msg = message_mod.buildHandshakeResp(&wire_buf, local_idx, hs_init.sender_index, ephemeral, msg2[keypair.key_size..]) catch return PeerManagerError.MessageError;

        // Get transport keys (swapped for responder)
        const keys = hs.split() catch return PeerManagerError.HandshakeFailed;

        // Create session
        const session = self.allocator.create(Session) catch return PeerManagerError.OutOfMemory;
        session.* = session_mod.Session.init(.{
            .local_index = local_idx,
            .remote_index = hs_init.sender_index,
            .send_key = keys.recv_cipher.key, // Swapped for responder
            .recv_key = keys.send_cipher.key,
            .remote_pk = initiator_pk,
        });

        // Register session
        {
            self.mutex.lock();
            defer self.mutex.unlock();
            peer.setSession(session);
            peer.setEndpoint(from);
            self.by_index.put(local_idx, peer) catch {};
        }

        // Send response
        self.transport.send(wire_msg, from) catch return PeerManagerError.TransportError;
    }
};

// =============================================================================
// Tests
// =============================================================================

const testing = std.testing;

test "new peer manager" {
    const kp = KeyPair.generate();
    const transport = try transport_mod.MockTransport.init(testing.allocator, "test");
    defer transport.deinit();

    var pm = PeerManager.init(testing.allocator, kp, Transport{ .mock = transport });
    defer pm.deinit();

    try testing.expectEqual(@as(usize, 0), pm.count());
}

test "add remove peer" {
    const kp = KeyPair.generate();
    const peer_kp = KeyPair.generate();
    const transport = try transport_mod.MockTransport.init(testing.allocator, "test");
    defer transport.deinit();

    var pm = PeerManager.init(testing.allocator, kp, Transport{ .mock = transport });
    defer pm.deinit();

    const peer = try testing.allocator.create(Peer);
    peer.* = Peer.init(testing.allocator, .{ .public_key = peer_kp.public });

    // Add peer
    try pm.addPeer(peer);
    try testing.expectEqual(@as(usize, 1), pm.count());

    // Try duplicate
    try testing.expectError(PeerManagerError.PeerExists, pm.addPeer(peer));

    // Get peer
    try testing.expect(pm.getPeer(peer_kp.public) != null);

    // Remove peer
    pm.removePeer(peer_kp.public);
    try testing.expectEqual(@as(usize, 0), pm.count());
    try testing.expect(pm.getPeer(peer_kp.public) == null);
}

test "send not established" {
    const kp = KeyPair.generate();
    const peer_kp = KeyPair.generate();
    const transport = try transport_mod.MockTransport.init(testing.allocator, "test");
    defer transport.deinit();

    var pm = PeerManager.init(testing.allocator, kp, Transport{ .mock = transport });
    defer pm.deinit();

    const peer = try testing.allocator.create(Peer);
    peer.* = Peer.init(testing.allocator, .{ .public_key = peer_kp.public });
    try pm.addPeer(peer);

    try testing.expectError(PeerManagerError.NotEstablished, pm.send(peer_kp.public, .chat, "test"));
}

test "send unknown peer" {
    const kp = KeyPair.generate();
    const unknown_kp = KeyPair.generate();
    const transport = try transport_mod.MockTransport.init(testing.allocator, "test");
    defer transport.deinit();

    var pm = PeerManager.init(testing.allocator, kp, Transport{ .mock = transport });
    defer pm.deinit();

    try testing.expectError(PeerManagerError.PeerNotFound, pm.send(unknown_kp.public, .chat, "test"));
}

test "clear" {
    const kp = KeyPair.generate();
    const transport = try transport_mod.MockTransport.init(testing.allocator, "test");
    defer transport.deinit();

    var pm = PeerManager.init(testing.allocator, kp, Transport{ .mock = transport });
    defer pm.deinit();

    for (0..5) |_| {
        const peer_kp = KeyPair.generate();
        const peer = try testing.allocator.create(Peer);
        peer.* = Peer.init(testing.allocator, .{ .public_key = peer_kp.public });
        try pm.addPeer(peer);
    }

    pm.clear();
    try testing.expectEqual(@as(usize, 0), pm.count());
}
